### DISCUSSION TITLE
word wrap in preview window

### DISCUSSION THREAD

**Discussion by korakinos** [junegunn/fzf#3383] [Q&A] [answered]
I'd like to have the option to get word wrap in the preview window (i.e. wrapping text on whitespace only). For prose text files, I would much prefer this over default behaviour of `--preview-window=wrap` to fill lines completely.

Is there a way to achieve this already? I couldn't find one.

If there isn't a way yet, I'm not sure which of these two features to request:

1. Word wrap support in fzf itself (e.g. `--preview-window=word-wrap`).
2. A way to inform preview commands about the current width of the preview window, possibly with a brace syntax like `{width}`, so that one could use a tool like `fmt` or `fold` to wrap a file: `--preview "fmt -w {width} {}"`.

**Comment by junegunn** [comment, marked as answer]
https://github.com/junegunn/fzf/blob/dd59b8c7b91082a4ae3a669979980abfaa3b470e/man/man1/fzf.1#L545-L549

**Reply by korakinos** [reply]
Perfect, thanks a lot! I should have read the manual more thoroughly.

**Comment by allefeld** [comment]
Access to the preview window width doesn't solve the word wrap problem generally, because as far as I can tell, none of the standard text word wrap filter programs – I looked at `fmt` and `fold` – can deal with ANSI color codes, leading to unnecessarily short lines if applied e.g. to colored grep results. Moreover, none have the ability to prefix wrapped lines like fzf does with `--wrap-sign=` (hanging indent).

So, it would imho still be great if fzf could implement word wrapping.

Another small quibble I have with fzf's wrapping is that scrolling moves by original lines, not wrapped lines. For very long lines, e.g. search results from Markdown files, this has a jarring effect.

I'm happy to create issues if that helps.

**Reply by junegunn** [reply]
Someone should be interested in writing a next-gen word wrap filter program that addresses the needs. Then it could be used with  fzf, but also independently, embracing the Unix philosophy.

**Reply by allefeld** [reply]
I'll take that as a "no". ;-)

I've searched for an external tool, but didn't find anything.
`less` can do it, but I wasn't able to access that functionality outside of displaying on screen. As soon as the output is piped, the wrapping is lost.
I've written a hacky version in bash which can deal with grep search results, but not a general solution.
A general solution could be based on the `ansiwrap` Python package, but that is unmaintained.

**Reply by junegunn** [reply]
I'm exploring alternative approaches and evaluating their merit. fzf embraces the Unix philosophy by delegating to other specialized programs (like `bat`, `ripgrep`, etc) as much as possible, but it's not clear this wrapping is a good fit for delegation. So I'm still undecided.

**Reply by acinis** [reply]
Based on this answer on [SE](https://unix.stackexchange.com/a/229549) (bottom part – fourth point), we can probably use `nroff`. It seems that Unicode and colors don't work out of the box (`preconv` can probably help with the former), but looking at what we all have on the `man` pages, it seems quite doable.

And all of these tools are already pre-installed on (most) *nix systems.

I ran into a similar problem a few weeks ago and I hope to find some free time to experiment (and share the results).

**Reply by junegunn** [reply]
Please test this: https://github.com/junegunn/fzf/commit/b6411beaa1ea2170fcdbb8b57bcb9522c189029f

**Reply by allefeld** [reply]
Seems to work correctly, thanks!

<img width="944" height="550" alt="image" src="https://github.com/user-attachments/assets/b1234351-1e03-4254-a264-36400c8039d4" />
